src/main/java/duke/storage/Storage.java

```diff
@@ -0,0 +1,196 @@
+                }
+            case 'E':
+                try {
+                    task = extractEvent(nextCommand.substring(7)); // [D][X] something at time
+                    break;
+                } catch (DataFileChangedException e) {
+                    throw new DataFileChangedException();
+                }
+            case 'T': // todos
+                task = new Todo(nextCommand.substring(7)); // disregards [T][X]
+                break;
+            default:
+                throw new DataFileChangedException();
+            }
+
+            // check if marked as done
+            if (nextCommand.charAt(4) == 'X') {
+                task.markAsDone();
+            }
+
+            taskList.add(task);
+        }
+
+        sc.close();
+
+        return taskList;
+    }
+
+    /**
+     * Extracts a deadline from the data file in the proper format for duke to read.
+     *
+     * @param text the deadline in the data file.
+     * @return a new deadline that represents the deadline from the data file.
+     * @throws DataFileChangedException if the data file was changed and any entry contains a wrong format.
+     */
+    private Deadline extractDeadline(String text) throws DataFileChangedException {
+        int lastOccurrenceOfBy = text.lastIndexOf(" (by: "); // in case other bys appear
+        String description = text.substring(0, lastOccurrenceOfBy);
+
+        // disregards "( by: " and trailing ")"
+        String by = text.substring(lastOccurrenceOfBy + 6, text.length() - 1);
+
+        LocalDateTime dateTime;
+
+        try {
+            dateTime = LocalDateTime.parse(by, DateTimeFormatter.ofPattern("MMM d yyyy, h:mm a"));
+        } catch (DateTimeParseException e) {
+            throw new DataFileChangedException();
+        }
+
+        return new Deadline(description, dateTime);
+    }
+
+    /**
+     * Extracts an event from the data file in the proper format for duke to read.
+     *
+     * @param text the event in the data file.
+     * @return a new deadline that represents the deadline from the data file.
+     * @throws DataFileChangedException if the data file was changed and any entry contains a wrong format.
+     */
+    private Event extractEvent(String text) throws DataFileChangedException {
+        int lastOccurrenceOfAt = text.lastIndexOf(" (at: ");
+        String description = text.substring(0, lastOccurrenceOfAt);
+
+        // disregards "( at: " and trailing ")"
+        String at = text.substring(lastOccurrenceOfAt + 6, text.length() - 1);
+
+        // prepare variables
+        String atWithoutWhiteSpace = at.replaceAll("\\s", "");
+        int lengthOfAtNoSpace = atWithoutWhiteSpace.length();
+
+        // throws error if it doesn't even contain sufficient number of characters for correct format
+        if (lengthOfAtNoSpace < 22 || lengthOfAtNoSpace > 25) { // MMM d yyyy, HH:mm - HH:mm
+            throw new DataFileChangedException();
+        }
```

Comment:
It seems like the parsing of your input string takes up a chunk of your code. I think you can consider abstracting out all these details into a new StringParser class (or push it into the Task classes). 

Maybe doing so might make the code more reader-friendly?